File: packages/app-ios/Sources/RunweaveIOS/Contracts/MobileLogin.swift
import Darwin
import Foundation

struct MobileLoginFailure: LocalizedError {
  let message: String
  var errorDescription: String? { message }
}

/// Mirrors @runweave/shared/mobile-login. The QR is a short-lived credential, never a log field.
struct MobileLoginQR: Decodable {
  let kind: String
  let version: Int
  let baseUrl: String
  let connectionName: String
  let requestId: String
  let qrSecret: String
  let expiresAt: String

  var expiry: Date? {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter.date(from: expiresAt) ?? ISO8601DateFormatter().date(from: expiresAt)
  }

  static func parse(_ text: String) throws -> Self {
    guard text.utf8.count <= 4096, let data = text.data(using: .utf8),
      let qr = try? JSONDecoder().decode(Self.self, from: data),
      qr.kind == "runweave.mobile-login"
    else { throw MobileLoginFailure(message: "这不是 Runweave 登录二维码，请重新扫码。") }
    guard qr.version == 1 else { throw MobileLoginFailure(message: "二维码版本不支持，请更新 App。") }
    guard UUID(uuidString: qr.requestId) != nil, validSecret(qr.qrSecret),
      !qr.connectionName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
      qr.connectionName.utf16.count <= 128, qr.expiresAt.utf8.count <= 64, qr.expiry != nil
    else { throw MobileLoginFailure(message: "二维码内容无效，请重新扫码。") }
    _ = try validatedBase(qr.baseUrl)
    // A phone clock cannot veto the server's authoritative expiry.
    return qr
  }

  static func validSecret(_ value: String) -> Bool {
    value.utf8.count == 43 && value.range(of: "^[A-Za-z0-9_-]{43}$", options: .regularExpression) != nil
  }

  static func validatedBase(_ value: String) throws -> URL {
    guard value.utf8.count <= 2048, !value.contains("\\"),
      value.rangeOfCharacter(from: .whitespacesAndNewlines) == nil,
      let parts = URLComponents(string: value), parts.query == nil, parts.fragment == nil,
      parts.user == nil, parts.password == nil, let rawHost = parts.host
    else { throw MobileLoginFailure(message: "二维码中的连接地址无效。") }
    let host = rawHost.lowercased().trimmingCharacters(in: CharacterSet(charactersIn: "[]."))
    guard host != "localhost", !host.hasSuffix(".localhost"), !host.contains("%") else {
      throw MobileLoginFailure(message: "请使用手机可访问的电脑地址，不能使用本机回环地址。")
    }
    var ipv4 = in_addr()
    if inet_aton(host, &ipv4) == 1 {
      let address = UInt32(bigEndian: ipv4.s_addr)
      guard address != 0, address >> 24 != 127 else { throw APIError.invalidURL }
    }
    var ipv6 = in6_addr()
    if inet_pton(AF_INET6, host, &ipv6) == 1 {
      let bytes = withUnsafeBytes(of: &ipv6) { Array($0) }
      let zeroPrefix = bytes.prefix(15).allSatisfy { $0 == 0 }
      let mapped = bytes.prefix(10).allSatisfy { $0 == 0 } && bytes[10] == 255 && bytes[11] == 255
      guard !(zeroPrefix && bytes[15] <= 1),
        !(mapped && (bytes[12] == 127 || bytes.suffix(4).allSatisfy { $0 == 0 }))
      else { throw APIError.invalidURL }
    }
    return try APIClient.normalize(value)
  }
}

enum MobileLoginState: String, Decodable {
  case waitingScan = "waiting_scan", pendingApproval = "pending_approval"
  case approved, issuing, issued, completed, cancelled, rejected, expired, unconfirmed, failed
}
struct MobileLoginStatus: Decodable {
  let state: MobileLoginState
  let expiresAt: String
  let claimId: String?
}
struct MobileLoginResult: Decodable {
  let accessToken: String
  let refreshToken: String
  let sessionId: String
  let expiresIn: Double

  func validate() throws {
    guard !accessToken.isEmpty, !refreshToken.isEmpty, accessToken.utf8.count <= 16384,
      refreshToken.utf8.count <= 16384, UUID(uuidString: sessionId) != nil,
      expiresIn.isFinite, expiresIn > 0
    else { throw APIError.invalidResponse }
  }
}

File: packages/shared/src/auth/mobile-login.ts
import type { LoginResponse } from "../protocol";

export const MOBILE_LOGIN_TTL_MS = 180_000;
export const MOBILE_LOGIN_RESULT_TTL_MS = 120_000;
export const MOBILE_LOGIN_MAX_QR_BYTES = 4096;

export interface MobileLoginQrV1 {
  kind: "runweave.mobile-login";
  version: 1;
  baseUrl: string;
  connectionName: string;
  requestId: string;
  qrSecret: string;
  expiresAt: string;
}

export type MobileLoginState =
  | "waiting_scan" | "pending_approval" | "approved" | "issuing"
  | "issued" | "completed" | "cancelled" | "rejected" | "expired"
  | "unconfirmed" | "failed";

export interface MobileLoginStatus {
  requestId: string;
  state: MobileLoginState;
  expiresAt: string;
  claimId?: string;
  deviceName?: string;
}

export interface CreateMobileLoginRequest { baseUrl: string; connectionName: string }
export interface ClaimMobileLoginRequest {
  qrSecret: string;
  claimantToken: string;
  deviceName: string;
  connectionId: string;
}
export interface MobileLoginClaim {
  claimId: string;
  state: MobileLoginState;
  expiresAt: string;
}
export interface MobileLoginDecision {
  claimId: string;
  decision: "approve" | "reject";
}
export type MobileLoginResult = LoginResponse & { refreshToken: string };
export type MobileLoginErrorCode =
  | "invalid_request" | "unauthorized" | "not_found" | "expired"
  | "conflict" | "already_issued" | "rate_limited" | "capacity"
  | "issuance_failed";
export interface MobileLoginFailure { code: MobileLoginErrorCode; message: string }

/** Shareable HTTP base, including a reverse proxy prefix. Never infer a LAN host. */
export function normalizeMobileLoginBaseUrl(raw: string): string {
  if (new TextEncoder().encode(raw).length > 2048 || raw !== raw.trim() || /[\\\s]/u.test(raw)) {
    throw new Error("Invalid mobile connection address");
  }
  const url = new URL(raw);
  const host = url.hostname.toLowerCase().replace(/\.$/, "");
  if (!/^https?:$/.test(url.protocol) || url.username || url.password ||
    raw.includes("?") || raw.includes("#") || !host ||
    host === "localhost" || host.endsWith(".localhost") ||
    /^127\./.test(host) || host === "0.0.0.0" ||
    host === "[::]" || host === "[::1]" ||
    /^\[::ffff:(?:7f[0-9a-f]{2}:|0:0\])/i.test(host)) {
    throw new Error("Configure a phone-accessible HTTP(S) connection address");
  }
  url.pathname = url.pathname.replace(/\/+$/, "");
  return url.toString().replace(/\/$/, "");
}
